Convert to validate pkg

thockin opened the pull request:
Start chipping away at the tangled web that is API validation.  This introduces a new "validate" package which can become the home for any non-API-type specific validator functions, all with the same naming and calling conventions.  This leaves callers saying things like:

`errs = append(errs, validate.DNSLabel(obj.Field, fldPath.Child("field"))...)`

The goal being to provide util validators for most common validations, even things like numeric comparisons, so we have consistent error messages and types.

xref #56230

```release-note
NONE
```

thockin committed:
Clean up validation utility functions

Creates a new package k8s.io/apimachinery/pkg/api/validate to replace
k8s.io/apimachinery/pkg/api/validation, and converts just one function -
validating non-negative integers.

thockin committed:
Convert validation to use new util package

Convert obvious validation sites doing manual non-negative checks.

thockin committed:
Add validate.NonNegativeQuantity for APIs

Convert callers of older function, remove.

thockin committed:
Simplify name validation

Get rid of the prefix logic, put it at the caller since we have very few
callers that need it.

Add a utility function to convert name validators into ErrorList
validators.

thockin committed:
Move DNS validators to util package

thockin committed:
Move PositiveQuantity to util pkg

thockin committed:
Add util func for validate.Positive()

thockin committed:
Replace ValidateImmutableAnnotation

thockin committed:
Add NonNegativeDuration validator

thockin committed:
Add validate.Immutable

thockin committed:
Add validate.InRange

thockin commented:
/retest

thockin commented:
For history's sake: everything up to InRange() is now green :)  I may add more commits.

thockin committed:
Add validate.IPString

php-coder commented:
Consider converting https://github.com/kubernetes/kubernetes/blob/f0e337cd56f36a6c9b8a0107084baa18fc20ddd8/pkg/apis/extensions/validation/validation.go#L877-L882 to use `validate.NonNegative()`

php-coder commented on staging/src/k8s.io/apimachinery/pkg/api/validate/validate.go:
```diff
@@ -0,0 +1,118 @@
+		allErrs = append(allErrs, field.Invalid(fldPath, value, `must be greater than or equal to 0`))
+	}
+	return allErrs
+}
```
Consider replacing `NonNegativeDuration` body by invocation of `NonNegative(int64(value), fldPath)`.

Or at least, I'd replace `must be greater than or equal to 0` by `isNegativeErrorMsg` constant.

yliaog commented:
/cc @mbohlool

errordeveloper commented:
@thockin have you tried rebasing this yet?

thockin committed:
Fix godoc comment

thockin committed:
Eliminate IsValidPathSegmentPrefix

As far as I can tell, the only reason this is here is pedantry.  I don't
think it's realistic to expect a `generateName: "."` to do anything
sane, especially since I don't know of any resource that supports
leading . in the name.

This makes follow cleanups easier.

thockin committed:
Use apimachinery validation directly
